internal/api/cluster.go

```diff
@@ -576,3 +582,86 @@ func handleGetAllUtilityMetadata(c *Context, w http.ResponseWriter, r *http.Requ
 	w.WriteHeader(http.StatusOK)
 	outputJSON(c, w, cluster.UtilityMetadata)
 }
+
+// handleAddClusterAnnotations responds to POST /api/cluster/{cluster}/annotations,
+// adds the set of annotations to the Cluster.
+func handleAddClusterAnnotations(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	clusterID := vars["cluster"]
+	c.Logger = c.Logger.WithField("cluster", clusterID).WithField("action", "add-cluster-annotations")
+
+	clusterDTO, status, unlockOnce := lockCluster(c, clusterID)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
+	annotations, err := annotationsFromRequest(r)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to get annotations from request")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	annotations, err = c.Store.CreateClusterAnnotations(clusterID, annotations)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to create cluster annotations")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	clusterDTO.Annotations = append(clusterDTO.Annotations, annotations...)
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	outputJSON(c, w, clusterDTO)
+}
+
+// handleDeleteClusterAnnotation responds to DELETE /api/cluster/{cluster}/annotation/{annotation-name},
+// removes annotation from the Cluster.
+func handleDeleteClusterAnnotation(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	clusterID := vars["cluster"]
+	annotationName := vars["annotation-name"]
+	c.Logger = c.Logger.
+		WithField("cluster", clusterID).
+		WithField("action", "delete-cluster-annotation").
```

Comment:
Is there a reason we don't allow for deleting multiple annotations at the same time? I feel like it is a bit strange that we can create multiple annotations with one call, but have to delete them one-by-one. What do you think?

Reply:
That results from API design. Following the REST convention, I am not a huge fan of passing body with the DELETE method therefor when deleting, the Annotation name is passed as a path variable.
GitHub API uses a similar approach with labels:
https://developer.github.com/v3/issues/labels/#remove-a-label-from-an-issue
https://developer.github.com/v3/issues/labels/#add-labels-to-an-issue

Reply:
Okay this makes sense. Thanks for your reasoning.

In the future we may want to standardize a bit with our other API update endpoints, but we can discuss this as needed.